### PR TITLE
Print per-second summary statistics detailing a scan, similar to ZMap

### LINKED ISSUE #472: [Feature Request] Periodically log statistics
**Describe the feature requested**

It would be useful to log statistics to stderr on some interval, similar to what zmap does:
```
0:02 25%; send: 1 done (444 p/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
```

I sometimes use `pv` for this:
```bash
cat domains.txt | zdns A | pv -l > results.json
```

but this is not perfect:
- Requires an additional tool
- Additional CPU usage
- Only counts lines. No statistics on good/bad queries.
- Output designed for interactive terminal usage, not logs.

Some statistics that zdns could log:
- Queries per second
- Success rate
  - Breakdown by DNS code?
  - Breakdown by resolver?

### PR DESCRIPTION
**Pull request by phillip-stephens**
Closes #472 

Adds a per-second summary with the following:
- h/min/sec printout
- \# of domains scanned
- domains/second
- success rate of domain resolution
- Occurrence of the different status codes, sorted greatest to least

A final summary is printed on scan completion

This is printed to stderr by default.

Two new CLI flags are available to modify this behavior:
```
  -q, --quiet                  do not print status updates
  -u, --status-updates-file=   file to write scan progress to, defaults to stderr (default: -)
```

## Demo
```
$ cat status
00h:00m:01s; 67 domains scanned; 66.99 domains/sec.; 100.0% success rate; NOERROR: 67
00h:00m:02s; 91 domains scanned; 45.49 domains/sec.; 100.0% success rate; NOERROR: 91
00h:00m:03s; 96 domains scanned; 32.00 domains/sec.; 100.0% success rate; NOERROR: 96
00h:00m:04s; 97 domains scanned; 24.24 domains/sec.; 100.0% success rate; NOERROR: 97
00h:00m:05s; 98 domains scanned; 19.60 domains/sec.; 100.0% success rate; NOERROR: 98
00h:00m:06s; 98 domains scanned; 16.33 domains/sec.; 100.0% success rate; NOERROR: 98
00h:00m:07s; 98 domains scanned; 14.00 domains/sec.; 100.0% success rate; NOERROR: 98
00h:00m:08s; 98 domains scanned; 12.25 domains/sec.; 100.0% success rate; NOERROR: 98
00h:00m:09s; 99 domains scanned; 11.00 domains/sec.; 99.0% success rate; NOERROR: 98, ITERATIVE_TIMEOUT: 1
00h:00m:10s; 99 domains scanned; 9.90 domains/sec.; 99.0% success rate; NOERROR: 98, ITERATIVE_TIMEOUT: 1
00h:00m:11s; 99 domains scanned; 9.00 domains/sec.; 99.0% success rate; NOERROR: 98, ITERATIVE_TIMEOUT: 1
00h:00m:12s; 99 domains scanned; 8.25 domains/sec.; 99.0% success rate; NOERROR: 98, ITERATIVE_TIMEOUT: 1
00h:00m:13s; 100 domains scanned; 7.69 domains/sec.; 99.0% success rate; NOERROR: 99, ITERATIVE_TIMEOUT: 1
00h:00m:13s; Scan Complete, no more input. 100 domains scanned; 7.18 domains/sec.; 99.0% success rate; NOERROR: 99, ITERATIVE_TIMEOUT: 1
```

### CODE REVIEW

**Review by zakird — CHANGES_REQUESTED** [review]
I don't think it's OK to make this not disable'able or not redirectable. I also don't think that it's OK to put these on `stdout` since it precludes piping ZDNS into another program. These should be able to be turned on and off and should be printed to stderr or a file. 

The most useful thing in ZMap to make sure a scan looks good is hitrate, which you can't actually quite get here. Can we get the same thing here. When calculating hitrate I would include both `NOERROR` and `NXDOMAIN` because both are a successful response from the DNS server.

**Comment by phillip-stephens** [comment]
Yep all seems reasonable! Will make those changes

File: src/cli/iohandlers/status_handler.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+		scanStartTime:   time.Now(),
+	}
+	ticker := time.NewTicker(time.Second)
+statusLoop:
+	for {
+		select {
+		case <-ticker.C:
+			// print per-second summary
+			timeSinceStart := time.Since(stats.scanStartTime)
+			s := fmt.Sprintf("%02dh:%02dm:%02ds; %d domains scanned; %.02f domains/sec.; %.01f%% success rate; %s\n",
```
Comment (zakird) [inline src/cli/iohandlers/status_handler.go]: I think that should be `names` not `domains`. I also think you could get rid of the `.` after sec.

File: src/cli/iohandlers/status_handler.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+			}
+			if _, ok = stats.statusOccurance[status]; !ok {
+				// initialize status if not seen before
+				stats.statusOccurance[status] = 0
+			}
+			stats.statusOccurance[status] += 1
+		}
+	}
+	timeSinceStart := time.Since(stats.scanStartTime)
+	s := fmt.Sprintf("%02dh:%02dm:%02ds; Scan Complete, no more input. %d domains scanned; %.02f domains/sec.; %.01f%% success rate; %s\n",
```
Comment (zakird) [inline src/cli/iohandlers/status_handler.go]: I think you can drop `, no more input`.  Would switch `domain` to `name` here.

**Review by zakird — APPROVED** [review]

**Pull request merged**